File: app/src/main/java/com/mx/gillustrated/component/MainActivityListView.java
package com.mx.gillustrated.component;

import android.content.Intent;
import android.os.Handler;
import android.os.Message;
import android.util.Log;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ListView;
import android.widget.RelativeLayout;
import android.widget.Toast;

import com.mx.gillustrated.R;
import com.mx.gillustrated.activity.DetailActivity;
import com.mx.gillustrated.activity.MainActivity;
import com.mx.gillustrated.adapter.DataListAdapter;
import com.mx.gillustrated.database.DataBaseHelper;
import com.mx.gillustrated.listener.ListenerListViewScrollHandler;
import com.mx.gillustrated.util.CommonUtil;
import com.mx.gillustrated.vo.CardInfo;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Created by maoxin on 2018/8/2.
 */

public class MainActivityListView {
    private static String INIT_ORDER_BY = CardInfo.ID;
    private static String INIT_ORDER_TYPE = CardInfo.SORT_DESC;

    private MainActivity mContext;
    private DataBaseHelper mOrmHelper;
    private DataView mDataView;
    private DataViewHandle mDataViewHandle;
    private List<CardInfo> mList;
    private DataListAdapter mAdapter;
    private int mListViewLastPosition; //保存最后一次本页面滚动位置

    private static final int  PAGE_SIZE = 20;
    private int currentPage;
    private int initPage; //初始页数，默认是1  如果从详细页面返回，可能为1+
    private int totalItemCount;

    private CardInfo mSearchCondition;
    private String mOrderBy;
    private String mIsAsc;


    public MainActivityListView(MainActivity context, DataBaseHelper ormHelper, DataViewHandle handle){
        mContext = context;
        mOrmHelper = ormHelper;
        mDataViewHandle = handle;
        mDataView = new DataView();
        initialize();
    }

    private void initialize(){
        mListViewLastPosition = mContext.getIntent().getIntExtra("position", 0);
        currentPage = initPage = mContext.getIntent().getIntExtra("currentPage", 1);
        String order =  mContext.getIntent().getStringExtra("orderBy");
        mOrderBy = order == null ? INIT_ORDER_BY : order.split("\\*")[0];
        mIsAsc =  order == null ? INIT_ORDER_TYPE : order.split("\\*")[1];

        mDataView.pageVboxLayout.setVisibility(android.view.View.GONE);
        mDataView.listViewMain.setOnItemClickListener(itemClickListener);
        mDataView.listViewMain.setOnScrollListener(new ListenerListViewScrollHandler( mDataView.listViewMain,  mDataView.pageVboxLayout, 0,
                new ListenerListViewScrollHandler.ScrollHandle(){
                    @Override
                    public void scrollLastRow(int totalCount) {
                        if( totalCount == totalItemCount  )
                            return;
                        currentPage++;
                        search();
                    }
                }));
        mList = new ArrayList();
        mAdapter = new DataListAdapter(mContext, mList);
    }


    public void search(){
        mDataViewHandle.onSearchStart();
        new Thread(new Runnable() {
            @Override
            public void run() {
                List<CardInfo> list =  mOrmHelper.getCardInfoDao().queryCards(mSearchCondition, mOrderBy, mIsAsc == CardInfo.SORT_ASC,
                        initPage == 1 ? (currentPage - 1) * PAGE_SIZE : 0, initPage * PAGE_SIZE );
                Message msg = Message.obtain();
                msg.obj = list;
                handler.sendMessage(msg);
            }
        }).start();
    }

    public void searchData(final CardInfo info){
        mSearchCondition = info;
        currentPage = 1;
        search();
    }

    public void searchData(String pinyin){
        mSearchCondition.setPinyinName(pinyin);
        currentPage = 1;
        search();
    }

    public void searchDataOrderby(String order){
        if(mOrderBy.equals(order))
            mIsAsc = mIsAsc.equals(CardInfo.SORT_ASC) ? CardInfo.SORT_DESC : CardInfo.SORT_ASC;
        else
            mIsAsc = CardInfo.SORT_DESC;
        mOrderBy = order;
        currentPage = 1;
        search();
    }

    public void setOrderBy(String orderBy, String isAsc){
        mOrderBy = orderBy == null ? INIT_ORDER_BY : orderBy;
        mIsAsc = isAsc == null ? INIT_ORDER_TYPE : isAsc;
    }

    public String getOrderBy(){
        return mOrderBy + "*" + mIsAsc;
    }

    public CardInfo getSearchCondition(){
        return mSearchCondition;
    }

    public int[] getIdList(){
        int[] list = new int[mList.size()];
        for(int i = 0; i < mList.size(); i++)
            list[i] = mList.get(i).getId();
        return list;
    }

    public  List<CardInfo> getDataList(){
        return mList;
    }


    private void updateList(boolean flag) {
        if (flag)
            mDataView.listViewMain.setAdapter(mAdapter);
        else
            mAdapter.notifyDataSetChanged();

        if(mListViewLastPosition > 0 ){
            mDataView.listViewMain.setSelection(mListViewLastPosition);
            mListViewLastPosition = 0;
        }

    }

    AdapterView.OnItemClickListener itemClickListener = new AdapterView.OnItemClickListener() {

        @Override
        public void onItemClick(AdapterView<?> arg0, View arg1, int position,
                                long arg3) {
            if (position >= 0) {
                CardInfo info = (CardInfo) arg0.getItemAtPosition(position);
                mDataViewHandle.onListItemClick(info, position, arg0.getCount(), currentPage);
            }
        }
    };

    Handler handler = new Handler() {
        @Override
        public void handleMessage(Message msg) {
            super.handleMessage(msg);
            List<CardInfo> list = (List<CardInfo>) msg.obj;
            totalItemCount = list.size() == 0 ? 0 : list.get(0).getTotalCount();

            if(currentPage == 1 || initPage > 1){
                initPage = 1;//恢复 只有初始第一次可能 >1
                mList.clear();
                mList.addAll(list);
                updateList(true);
            }else{
                mList.addAll(list);
                updateList(false);
            }
            mDataViewHandle.onSearchCompleted();
        }
    };

    class DataView{
        @BindView(R.id.lvMain)
        ListView listViewMain;
        @BindView(R.id.pageVBox)
        RelativeLayout pageVboxLayout;

        DataView(){
            ButterKnife.bind(this, mContext);
        }

    }

    public interface DataViewHandle{
        void onListItemClick(CardInfo info, int position, int totalCount, int currentPage);
        void onSearchCompleted();
        void onSearchStart();
    }

}

File: app/src/main/java/com/mx/gillustrated/component/MainActivityListView.kt
package com.mx.gillustrated.component

import android.os.Handler
import android.os.Looper
import android.os.Message
import android.widget.AdapterView
import com.mx.gillustrated.activity.MainActivity
import com.mx.gillustrated.adapter.DataListAdapter
import com.mx.gillustrated.database.DataBaseHelper
import com.mx.gillustrated.listener.ListenerListViewScrollHandler
import com.mx.gillustrated.vo.CardInfo
import java.util.ArrayList
import com.mx.gillustrated.activity.BaseActivity.Companion.SHARE_PAGE_SIZE
import com.mx.gillustrated.databinding.ActivityMainBinding
import java.lang.ref.WeakReference

/**
 * Created by maoxin on 2018/8/2.
 */

class MainActivityListView(private val mContext: MainActivity, private val binding: ActivityMainBinding,
                           private val mOrmHelper: DataBaseHelper, private val mDataViewHandle: DataViewHandle) {

    private var mList: MutableList<CardInfo> = ArrayList()
    private var mAdapter: DataListAdapter? = null
    private var mListViewLastPosition: Int = 0 //保存最后一次本页面滚动位置
    private var currentPage: Int = 0
    private var initPage: Int = 0 //初始页数，默认是1  如果从详细页面返回，可能为1+
    private var totalItemCount: Int = 0
    private val listHandler:ListHandler = ListHandler(this)
    var searchCondition: CardInfo = CardInfo()
        private set
    private lateinit var mOrderBy: String
    private lateinit var mIsAsc: String
    val orderBy: String
        get() = "$mOrderBy*$mIsAsc"

    val idListWithProfile: IntArray
        get() {
            val temp =  mutableListOf<Int>()
            for (i in mList.indices) {
                if ("Y" == mList[i].profile)
                    temp.add(mList[i].id)
            }
            val result = IntArray(temp.size)
            for (i in temp.indices) {
                result[i] = temp[i]
            }
            return result
        }


    val dataList: List<CardInfo>
        get() = mList

    private val itemClickListener = AdapterView.OnItemClickListener { arg0, _, position, _ ->
        if (position >= 0) {
            val info = arg0.getItemAtPosition(position) as CardInfo
            mDataViewHandle.onListItemClick(info, position, arg0.count, currentPage)
        }
    }

    private class ListHandler constructor(component: MainActivityListView ) : Handler(Looper.getMainLooper()) {

        private val weakReference: WeakReference<MainActivityListView> = WeakReference(component)

        override fun handleMessage(msg: Message) {
            super.handleMessage(msg)
            val component = weakReference.get()!!
            @Suppress("UNCHECKED_CAST") val list = msg.obj as List<CardInfo>
            component.totalItemCount = if (list.isEmpty()) 0 else list[0].totalCount

            if (component.currentPage == 1 || component.initPage > 1) {
                component.initPage = 1//恢复 只有初始第一次可能 >1
                component.mList.clear()
                component.mList.addAll(list)
                component.updateList(true)
            } else {
                component.mList.addAll(list)
                component.updateList(false)
            }
            component.mDataViewHandle.onSearchCompleted()
        }
    }

    init {
        initialize()
    }

    private fun initialize() {
        mListViewLastPosition = mContext.intent.getIntExtra("position", 0)
        initPage = mContext.intent.getIntExtra("currentPage", 1)
        currentPage = initPage
        val order = mContext.intent.getStringExtra("orderBy")
        mOrderBy = if (order == null) INIT_ORDER_BY else order.split("\\*".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0]
        mIsAsc = if (order == null) INIT_ORDER_TYPE else order.split("\\*".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[1]

        binding.pageVBox.visibility = android.view.View.GONE
        binding.lvMain.onItemClickListener = itemClickListener
        binding.lvMain.setOnScrollListener(ListenerListViewScrollHandler(binding.lvMain, binding.pageVBox, 0,
                object : ListenerListViewScrollHandler.ScrollHandle {
                    override fun scrollLastRow(totalCount: Int) {
                        if (totalCount == totalItemCount)
                            return
                        currentPage++
                        search()
                    }
                }))
        mAdapter = DataListAdapter(mContext, mList)
    }


    private fun search() {
        mDataViewHandle.onSearchStart()
        val pageSize = mContext.mSP.getInt(SHARE_PAGE_SIZE + mContext.mGameType, PAGE_SIZE)
        Thread{
            val list = mOrmHelper.cardInfoDao.queryCards(searchCondition, mOrderBy, mIsAsc == CardInfo.SORT_ASC,
                    (if (initPage == 1) (currentPage - 1) * pageSize else 0).toLong(), (initPage * pageSize).toLong())
            val msg = Message.obtain()
            msg.obj = list
            listHandler.sendMessage(msg)
        }.start()
    }

    fun searchData(info: CardInfo) {
        searchCondition = info
        currentPage = 1
        search()
    }

    fun searchData(pinyin: String) {
        searchCondition.pinyinName = pinyin
        currentPage = 1
        search()
    }

    fun searchDataOrderBy(order: String) {
        mIsAsc = if (mOrderBy == order)
            if (mIsAsc == CardInfo.SORT_ASC) CardInfo.SORT_DESC else CardInfo.SORT_ASC
        else
            CardInfo.SORT_DESC
        mOrderBy = order
        currentPage = 1
        search()
    }

    fun setOrderBy(orderBy: String?, isAsc: String?) {
        mOrderBy = orderBy ?: INIT_ORDER_BY
        mIsAsc = isAsc ?: INIT_ORDER_TYPE
    }


    private fun updateList(flag: Boolean) {
        if (flag)
            binding.lvMain.adapter = mAdapter
        else
            mAdapter!!.notifyDataSetChanged()

        if (mListViewLastPosition > 0) {
            binding.lvMain.setSelection(mListViewLastPosition)
            mListViewLastPosition = 0
        }

    }


    interface DataViewHandle {
        fun onListItemClick(info: CardInfo, position: Int, totalCount: Int, currentPage: Int)
        fun onSearchCompleted()
        fun onSearchStart()
    }

    companion object {
        private const val INIT_ORDER_BY = CardInfo.ID
        private const val INIT_ORDER_TYPE = CardInfo.SORT_DESC
        private const val PAGE_SIZE = 50
    }

}
